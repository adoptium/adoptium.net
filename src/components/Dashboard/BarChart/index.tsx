"use client"

import React from 'react'
import Highcharts from 'highcharts'
import HighchartsReact from 'highcharts-react-official'
import startCase from 'lodash/startCase'

interface BarChartProps {
    data?: { [key: string]: number };
    name: string;
    startCaseKeys?: boolean;
}

const BarChart: React.FC<BarChartProps> = ({ data, name, startCaseKeys }) => {
    if (!data) return null

    const options = {
        chart: {
            type: 'column',
            backgroundColor: 'transparent',
            style: { fontFamily: 'inherit' },
            animation: process.env.NODE_ENV !== "test"
        },
        colors: [
            "#2caffe", "#6c6de3", "#feae62", "#a259f7", "#f67280", "#43e97b"
        ],
        accessibility: {
            enabled: false,
        },
        title: {
            text: name,
            style: { color: '#fff', fontWeight: 'bold', fontSize: '1.25rem' }
        },
        subtitle: {
            text: 'Data is from: <a href="https://api.adoptium.net/v3/stats/downloads/total" target="_blank" rel="noopener noreferrer">api.adoptium.net/v3/stats/downloads/total</a>',
            useHTML: true,
            style: { color: '#bdbdbd' }
        },
        xAxis: {
            categories: Object.keys(data).map(v => startCaseKeys ? startCase(v) : v),
            crosshair: true,
            labels: { style: { color: '#fff', fontWeight: 'bold', fontSize: '1rem' } },
            lineColor: 'rgba(255,255,255,0.2)',
            tickColor: 'rgba(255,255,255,0.2)',
        },
        yAxis: {
            title: {
                text: 'Downloads',
                style: { color: '#fff', fontWeight: 'bold', fontSize: '1rem' }
            },
            labels: { style: { color: '#fff', fontWeight: 'bold', fontSize: '1rem' } },
            gridLineColor: 'rgba(255,255,255,0.1)'
        },
        legend: {
            itemStyle: { color: '#fff', fontWeight: 'bold' },
            itemHoverStyle: { color: '#6ba4e7' },
        },
        tooltip: {
            backgroundColor: '#2c225a',
            borderColor: '#6ba4e7',
            style: { color: '#fff', fontWeight: 'bold' },
            headerFormat: '<span style="font-size:10px">{point.key}</span><table>',
            footerFormat: '</table>',
            shared: true,
            useHTML: true
        },
        plotOptions: {
            column: {
                borderRadius: 6,
                dataLabels: {
                    enabled: true,
                    style: { color: '#fff', fontWeight: 'bold', textOutline: 'none' },
                    formatter: function (this: Highcharts.Point) {
                        if (!this || typeof this.y !== 'number') return '';
                        return Highcharts.numberFormat(this.y, 0, '.', ' ');
                    }
                },
                pointPadding: 0.2,
                borderWidth: 0,
                minPointLength: 10,
                shadow: false
            }
        },
        series: [{
            name,
            data: Object.values(data)
        }]
    }

    return (
        <div className='chart'>
            <HighchartsReact
                highcharts={Highcharts}
                options={options}
            />
        </div>
    )
}

export default BarChart
