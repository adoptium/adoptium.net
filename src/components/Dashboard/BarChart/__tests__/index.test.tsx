import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { describe, it, expect, afterEach } from 'vitest';
import BarChart from '../index';

describe('BarChart', () => {
    afterEach(() => {
        cleanup();
    });

    it('renders nothing if no data', () => {
        const { container } = render(<BarChart name="Test Chart" />);
        expect(container.firstChild).toBeNull();
    });
});
