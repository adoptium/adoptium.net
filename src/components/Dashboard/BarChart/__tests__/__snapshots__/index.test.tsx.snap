// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BarChart > matches snapshot 1`] = `
<div
  class="chart"
>
  <div
    data-highcharts-chart="0"
    style="overflow: hidden;"
  >
    <div
      class="highcharts-container "
      dir="ltr"
      style="position: relative; overflow: hidden; width: 600px; height: 400px; text-align: left; line-height: normal; z-index: 0; user-select: none; outline: none; padding: 0px; font-family: inherit;"
    >
      <svg
        aria-label="Test Chart"
        class="highcharts-root"
        height="400"
        role="img"
        style="font-family: inherit; font-size: 1rem;"
        version="1.1"
        viewBox="0 0 600 400"
        width="600"
        xmlns="http://www.w3.org/2000/svg"
      >
        <desc>
          Created with Highcharts 12.3.0
        </desc>
        <defs>
          <filter>
            <fedropshadow
              dx="1"
              dy="1"
              flood-color="#000000"
              flood-opacity="0.75"
              stdDeviation="2.5"
            />
          </filter>
          <clippath>
            <rect
              fill="none"
              height="NaN"
              width="NaN"
              x="0"
              y="0"
            />
          </clippath>
        </defs>
        <rect
          class="highcharts-background"
          fill="transparent"
          filter="none"
          height="400"
          rx="0"
          ry="0"
          width="600"
          x="0"
          y="0"
        />
        <rect
          class="highcharts-plot-background"
          fill="none"
          filter="none"
          height="NaN"
          width="NaN"
          x="NaN"
          y="10"
        />
        <g
          class="highcharts-grid highcharts-xaxis-grid"
          data-z-index="1"
        >
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke="#e6e6e6"
            stroke-dasharray="none"
            stroke-width="0"
          />
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke="#e6e6e6"
            stroke-dasharray="none"
            stroke-width="0"
          />
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke="#e6e6e6"
            stroke-dasharray="none"
            stroke-width="0"
          />
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke="#e6e6e6"
            stroke-dasharray="none"
            stroke-width="0"
          />
        </g>
        <g
          class="highcharts-grid highcharts-yaxis-grid"
          data-z-index="1"
        />
        <rect
          class="highcharts-plot-border"
          data-z-index="1"
          fill="none"
          height="0"
          stroke="#cccccc"
          stroke-width="0"
          width="0"
          x="0"
          y="10"
        />
        <g
          class="highcharts-axis highcharts-xaxis"
          data-z-index="2"
        >
          <path
            class="highcharts-axis-line"
            d="M 0 0"
            data-z-index="7"
            fill="none"
            stroke="rgba(255,255,255,0.2)"
            stroke-width="1"
          />
        </g>
        <g
          class="highcharts-axis highcharts-yaxis"
          data-z-index="2"
        >
          <text
            class="highcharts-axis-title"
            data-z-index="7"
            style="font-size: 1rem; font-weight: bold; fill: #fff;"
            text-anchor="middle"
            transform="translate(0,0) rotate(270 NaN NaN)"
            x="NaN"
            y="NaN"
          >
            Downloads
          </text>
          <path
            class="highcharts-axis-line"
            d="M 0 0"
            data-z-index="7"
            fill="none"
            stroke="#333333"
            stroke-width="0"
          />
        </g>
        <g
          class="highcharts-series-group"
          data-z-index="3"
          filter="none"
        >
          <g
            class="highcharts-series highcharts-series-0 highcharts-column-series highcharts-color-0 highcharts-tracker"
            data-z-index="0.1"
            opacity="1"
            transform="translate(NaN,10) scale(1 0.001)"
          />
          <g
            class="highcharts-markers highcharts-series-0 highcharts-column-series highcharts-color-0"
            data-z-index="0.1"
            opacity="1"
            transform="translate(NaN,10) scale(1 1)"
          />
        </g>
        <text
          class="highcharts-title"
          data-z-index="4"
          style="font-size: 1.25rem; font-weight: bold; fill: #fff;"
          text-align="center"
          text-anchor="middle"
          transform="translate(0,0) scale(1 1)"
          transform-origin="10 4"
          x="300"
          y="10"
        >
          Test Chart
        </text>
        <text
          class="highcharts-caption"
          data-z-index="4"
          style="font-size: 0.8em; fill: #666666;"
          text-align="left"
          text-anchor="start"
          transform="translate(0,0) scale(1 1)"
          transform-origin="10 3"
          x="10"
          y="387"
        />
        <g
          class="highcharts-datalabels-group"
          data-z-index="6"
        >
          <g
            class="highcharts-data-labels highcharts-series-0 highcharts-column-series highcharts-color-0 highcharts-tracker"
            data-z-index="6"
            opacity="0"
            transform="translate(NaN,10) scale(1 1)"
          >
            <g
              class="highcharts-label highcharts-data-label highcharts-data-label-color-0"
              data-z-index="1"
              filter="none"
              transform="translate(0,0)"
              visibility="hidden"
            >
              <text
                data-z-index="1"
                style="font-size: 0.7em; font-weight: bold; fill: #fff;"
                x="NaN"
                y="7"
              >
                1 234
              </text>
            </g>
            <g
              class="highcharts-label highcharts-data-label highcharts-data-label-color-0"
              data-z-index="1"
              filter="none"
              transform="translate(0,0)"
              visibility="hidden"
            >
              <text
                data-z-index="1"
                style="font-size: 0.7em; font-weight: bold; fill: #fff;"
                x="NaN"
                y="7"
              >
                5 678
              </text>
            </g>
            <g
              class="highcharts-label highcharts-data-label highcharts-data-label-color-0"
              data-z-index="1"
              filter="none"
              transform="translate(0,0)"
              visibility="hidden"
            >
              <text
                data-z-index="1"
                style="font-size: 0.7em; font-weight: bold; fill: #fff;"
                x="NaN"
                y="7"
              >
                6 912
              </text>
            </g>
          </g>
        </g>
        <g
          class="highcharts-legend highcharts-no-tooltip"
          data-z-index="7"
          text-align="center"
          transform="translate(300,367)"
        >
          <rect
            class="highcharts-legend-box"
            fill="none"
            filter="none"
            rx="0"
            ry="0"
            stroke="#999999"
            stroke-width="0"
          />
          <g
            data-z-index="1"
          >
            <g>
              <g
                class="highcharts-legend-item highcharts-column-series highcharts-color-0 highcharts-series-0"
                data-z-index="1"
                transform="translate(8,3)"
              >
                <text
                  data-z-index="2"
                  style="cursor: pointer; font-size: 0.8em; text-decoration: none; font-weight: bold; fill: #fff;"
                  text-anchor="start"
                  x="21"
                  y="5"
                >
                  Test Chart
                </text>
                <rect
                  class="highcharts-point"
                  data-z-index="3"
                  fill="#2caffe"
                  height="0"
                  rx="0"
                  ry="0"
                  width="0"
                  x="8"
                  y="6"
                />
              </g>
            </g>
          </g>
        </g>
        <g
          class="highcharts-axis-labels highcharts-xaxis-labels"
          data-z-index="7"
        >
          <text
            style="cursor: default; font-size: 1rem; font-weight: bold; fill: #fff;"
            x="0"
          >
            x
          </text>
          <text
            style="cursor: default; font-size: 1rem; font-weight: bold; fill: #fff;"
            text-anchor="middle"
            transform="translate(0,0)"
            x="0"
          >
            github
          </text>
          <text
            style="cursor: default; font-size: 1rem; font-weight: bold; fill: #fff;"
            text-anchor="middle"
            transform="translate(0,0)"
            x="0"
          >
            dockerhub
          </text>
          <text
            style="cursor: default; font-size: 1rem; font-weight: bold; fill: #fff;"
            text-anchor="middle"
            transform="translate(0,0)"
            x="0"
          >
            total
          </text>
        </g>
        <g
          class="highcharts-axis-labels highcharts-yaxis-labels"
          data-z-index="7"
        />
        <text
          class="highcharts-credits"
          data-z-index="8"
          style="cursor: pointer; font-size: 0.6em; fill: #999999;"
          text-align="right"
          text-anchor="end"
          x="590"
          y="395"
        >
          Highcharts.com
        </text>
      </svg>
      <span
        class="highcharts-subtitle"
        data-z-index="4"
        style="position: absolute; font-family: inherit; font-size: 0.8em; white-space: nowrap; text-overflow: ellipsis; overflow: hidden; margin-left: 0px; margin-top: 0px; left: 300px; top: 8px; text-align: center; transform-origin: 0px 2px; color: rgb(189, 189, 189); transform: scale(1,1);"
        text-align="center"
        transform-origin="10 3"
      >
        Data is from: 
        <a
          href="https://api.adoptium.net/v3/stats/downloads/total"
          target="_blank"
        >
          api.adoptium.net/v3/stats/downloads/total
        </a>
      </span>
    </div>
  </div>
</div>
`;
